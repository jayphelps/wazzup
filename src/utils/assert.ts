export const assert = (value: boolean, msg = 'Assertion failed') => {
  if (value === false) {
    throw new Error(msg);
  }
};
