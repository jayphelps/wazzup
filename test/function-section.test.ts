import { ast, wasm } from '../';
import { assertModule } from './utils/assertModule';

describe('Function Section', () => {
  let module;

  beforeEach(() => {
    module = new ast.Module();
  });

  test('adds functions', () => {
    const type1 = module.addFunctionType();

    module.addFunction('foo', type1, [], []);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (func (;0;) (type 0)))
    `);

    const type2 = module.addFunctionType();

    module.addFunction('bar', type2, [], []);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (type (;1;) (func))
      |   (func (;0;) (type 0))
      |   (func (;1;) (type 1)))
    `);
  });

  test('adds functions with local types', () => {
    const type1 = module.addFunctionType();
    const localTypes1 = [wasm.Type.i32, wasm.Type.i64];
    module.addFunction('foo', type1, localTypes1, []);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (func (;0;) (type 0)
      |     (local i32 i64)))
    `);

    const type2 = module.addFunctionType();
    const localTypes2 = [wasm.Type.i32, wasm.Type.i32, wasm.Type.i64];
    module.addFunction('bar', type2, localTypes2, []);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (type (;1;) (func))
      |   (func (;0;) (type 0)
      |     (local i32 i64))
      |   (func (;1;) (type 1)
      |     (local i32 i32 i64)))
    `);
  });
});
