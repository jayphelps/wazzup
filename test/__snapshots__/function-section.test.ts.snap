// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function Section adds functions 1`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Function Section adds functions 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  4,
  1,
  96,
  0,
  0,
  3,
  2,
  1,
  0,
  10,
  4,
  1,
  2,
  0,
  11,
]
`;

exports[`Function Section adds functions 3`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "bar",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Function Section adds functions 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  7,
  2,
  96,
  0,
  0,
  96,
  0,
  0,
  3,
  3,
  2,
  0,
  1,
  10,
  7,
  2,
  2,
  0,
  11,
  2,
  0,
  11,
]
`;

exports[`Function Section adds functions with local types 1`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [
        127,
        126,
      ],
      "name": "foo",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Function Section adds functions with local types 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  4,
  1,
  96,
  0,
  0,
  3,
  2,
  1,
  0,
  10,
  8,
  1,
  6,
  2,
  1,
  127,
  1,
  126,
  11,
]
`;

exports[`Function Section adds functions with local types 3`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [
        127,
        126,
      ],
      "name": "foo",
      "type": 112,
    },
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [
        127,
        127,
        126,
      ],
      "name": "bar",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Function Section adds functions with local types 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  7,
  2,
  96,
  0,
  0,
  96,
  0,
  0,
  3,
  3,
  2,
  0,
  1,
  10,
  15,
  2,
  6,
  2,
  1,
  127,
  1,
  126,
  11,
  6,
  2,
  2,
  127,
  1,
  126,
  11,
]
`;
